File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/VLM/RunAnywhere+VisionLanguage.swift
//
//  RunAnywhere+VisionLanguage.swift
//  RunAnywhere SDK
//
//  Public API for Vision Language Model (VLM) operations.
//  Uses C++ directly via CppBridge.VLM.
//
//  Loading flows through the canonical lifecycle (`RAModelLoadRequest`).
//  This file owns only inference (image processing) entry points.
//

import CRACommons

// C struct with raw pointers — safe to send across concurrency boundaries
// because the backing Data (rgbData) is kept alive alongside it.
// `@retroactive` acknowledges we're extending a type imported from CRACommons.
extension rac_vlm_image_t: @retroactive @unchecked Sendable {}

// MARK: - Vision Language Model

public extension RunAnywhere {

    /// Process a generated-proto VLM image through the C++ VLM ABI.
    static func processImage(
        _ image: RAVLMImage,
        options: RAVLMGenerationOptions
    ) async throws -> RAVLMResult {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }
        try await ensureServicesReady()

        // Query ModelLifecycle (the canonical source of truth) instead of
        // the CppBridge.VLM actor's per-handle state. VLM accepts both
        // `.multimodal` and `.vision` — try `.multimodal` first (the
        // canonical category used by SDK consumers and example apps), fall
        // back to `.vision` so models loaded under either category are
        // honored. Both collapse to `SDK_COMPONENT_VLM` in C++ commons.
        guard isVLMModelLoaded() else {
            throw SDKException(code: .notInitialized, message: "VLM model not loaded", category: .component)
        }

        return try await CppBridge.VLM.shared.process(image: image, options: options)
    }

    /// Stream typed VLM events from C++.
    ///
    /// Canonical cross-SDK shape: yields generated `RAVLMStreamEvent`s
    /// (STARTED → TOKEN* → exactly one terminal COMPLETED/ERROR; COMPLETED
    /// carries the full `RAVLMResult` with metrics). The prompt travels in
    /// `options.prompt`. Backed by `rac_vlm_stream_proto`, the same typed
    /// ABI Flutter and React Native consume.
    static func processImageStream(
        _ image: RAVLMImage,
        options: RAVLMGenerationOptions
    ) async throws -> AsyncStream<RAVLMStreamEvent> {
        guard isInitialized else {
            throw SDKException(code: .notInitialized, message: "SDK not initialized", category: .internal)
        }
        try await ensureServicesReady()

        guard isVLMModelLoaded() else {
            throw SDKException(code: .notInitialized, message: "VLM model not loaded", category: .component)
        }

        return try await CppBridge.VLM.shared.processStream(image: image, options: options)
    }

    /// Ergonomic overload mirroring React Native: the prompt is applied onto
    /// `options.prompt` before streaming.
    static func processImageStream(
        _ image: RAVLMImage,
        prompt: String,
        options: RAVLMGenerationOptions = .defaults()
    ) async throws -> AsyncStream<RAVLMStreamEvent> {
        var effectiveOptions = options
        effectiveOptions.prompt = prompt
        return try await processImageStream(image, options: effectiveOptions)
    }

    /// Returns true if a VLM model is loaded in the lifecycle under either
    /// the `.multimodal` or `.vision` category. Both categories collapse to
    /// `SDK_COMPONENT_VLM` in C++ commons.
    private static func isVLMModelLoaded() -> Bool {
        firstLoadedModelSnapshot(categories: [.multimodal, .vision]) != nil
    }

    /// Cancel the current VLM generation.
    static func cancelVLMGeneration() async {
        await CppBridge.VLM.shared.cancel()
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-web/packages/core/src/Public/Extensions/RunAnywhere+VisionLanguage.ts
/**
 * RunAnywhere+VisionLanguage.ts
 *
 * Public VLM namespace matching Swift's RunAnywhere+VisionLanguage. The Web
 * implementation delegates to a backend-installed provider so app code never
 * imports backend worker bridges directly.
 */

import type {
  VLMGenerationOptions,
  VLMImage,
  VLMResult,
  VLMStreamEvent,
} from '@runanywhere/proto-ts/vlm_options';
import {
  VLMGenerationOptions as VLMGenerationOptionsMessage,
  VLMModelFamily,
} from '@runanywhere/proto-ts/vlm_options';
import {
  ModelCategory,
  type CurrentModelResult,
} from '@runanywhere/proto-ts/model_types';
import { SDKException } from '../../Foundation/SDKException.js';
import { WebModelLifecycle } from './RunAnywhere+ModelLifecycle.js';

export interface VisionLanguageProvider {
  readonly isInitialized: boolean;
  readonly isModelLoaded: boolean;
  loadCurrentModel?(currentModel: CurrentModelResult): Promise<void>;
  unloadModel?(): Promise<void>;
  processImage(image: VLMImage, options: VLMGenerationOptions): Promise<VLMResult>;
  /** Typed stream: STARTED → TOKEN* → exactly one terminal COMPLETED/ERROR
   *  (COMPLETED carries the full VLMResult). Canonical cross-SDK shape. */
  processImageStream?(image: VLMImage, options: VLMGenerationOptions): Promise<AsyncIterable<VLMStreamEvent>>;
  cancelVLMGeneration(): Promise<void> | void;
}

let provider: VisionLanguageProvider | null = null;

export function setVisionLanguageProvider(next: VisionLanguageProvider | null): void {
  provider = next;
}

function requireProvider(feature: string): VisionLanguageProvider {
  if (provider) return provider;
  throw SDKException.backendNotAvailable(
    feature,
    'No Web vision-language provider is registered. Call LlamaCPP.register() first.',
  );
}

async function processImageStream(
  image: VLMImage,
  options: VLMGenerationOptions,
): Promise<AsyncIterable<VLMStreamEvent>>;
/**
 * Ergonomic overload mirroring Swift/React Native: the prompt is applied
 * onto `options.prompt` before streaming. When `options` is omitted, the
 * remaining knobs fall back to the canonical defaults
 * (`RAVLMGenerationOptions.defaults()` — applied by
 * `normalizeVLMGenerationOptions`). Swift parity:
 * `processImageStream(_:prompt:options:)` (RunAnywhere+VisionLanguage.swift:71-79).
 */
async function processImageStream(
  image: VLMImage,
  prompt: string,
  options?: VLMGenerationOptions,
): Promise<AsyncIterable<VLMStreamEvent>>;
async function processImageStream(
  image: VLMImage,
  optionsOrPrompt: VLMGenerationOptions | string,
  maybeOptions?: VLMGenerationOptions,
): Promise<AsyncIterable<VLMStreamEvent>> {
  const options = typeof optionsOrPrompt === 'string'
    ? { ...(maybeOptions ?? VLMGenerationOptionsMessage.fromPartial({})), prompt: optionsOrPrompt }
    : optionsOrPrompt;
  const active = requireProvider('visionLanguage.processImageStream');
  if (!active.processImageStream) {
    throw SDKException.backendNotAvailable(
      'visionLanguage.processImageStream',
      'The active Web vision-language provider does not expose streaming.',
    );
  }
  return active.processImageStream(image, normalizeVLMGenerationOptions(options, true));
}

export const VisionLanguage = {
  get isInitialized(): boolean {
    return provider?.isInitialized ?? false;
  },

  get isModelLoaded(): boolean {
    return provider?.isModelLoaded ?? false;
  },

  async loadCurrentModel(): Promise<void> {
    // Swift parity (RunAnywhere+VisionLanguage.swift): VLM accepts both
    // `.multimodal` and `.vision` — try `.multimodal` first (the canonical
    // category), fall back to `.vision`. No any-category fallback: a loaded
    // LLM/STT model must not satisfy the VLM guard.
    const current =
      WebModelLifecycle.currentModel({
        category: ModelCategory.MODEL_CATEGORY_MULTIMODAL,
        includeModelMetadata: true,
      }) ??
      WebModelLifecycle.currentModel({
        category: ModelCategory.MODEL_CATEGORY_VISION,
        includeModelMetadata: true,
      });

    if (!current?.modelId) {
      // Swift parity: RunAnywhere+VisionLanguage.swift:40 throws `.notInitialized` ("VLM model not loaded").
      throw SDKException.notInitialized(
        'No VLM model is loaded. Call RunAnywhere.loadModel(...) with a multimodal model before RunAnywhere.processImage().',
      );
    }

    const active = requireProvider('visionLanguage.loadCurrentModel');
    if (!active.loadCurrentModel) {
      throw SDKException.backendNotAvailable(
        'visionLanguage.loadCurrentModel',
        'The active Web vision-language provider cannot load C++ lifecycle resolved artifacts.',
      );
    }
    await active.loadCurrentModel(current);
  },

  async unloadModel(): Promise<void> {
    const active = requireProvider('visionLanguage.unloadModel');
    if (!active.unloadModel) return;
    await active.unloadModel();
  },

  processImage(image: VLMImage, options: VLMGenerationOptions): Promise<VLMResult> {
    return requireProvider('visionLanguage.processImage').processImage(
      image,
      normalizeVLMGenerationOptions(options, false),
    );
  },

  /**
   * Typed VLM event stream. Also accepts the `(image, prompt, options?)`
   * convenience overload — Swift parity:
   * `processImageStream(_:prompt:options:)` (RunAnywhere+VisionLanguage.swift:71-79).
   */
  processImageStream,

  async cancelVLMGeneration(): Promise<void> {
    await requireProvider('visionLanguage.cancelVLMGeneration').cancelVLMGeneration();
  },
};

export type VisionLanguageCapability = typeof VisionLanguage;

function normalizeVLMGenerationOptions(
  options: VLMGenerationOptions,
  streamingEnabled: boolean,
): VLMGenerationOptions {
  // Defaults mirror Swift `RAVLMGenerationOptions.defaults()`
  // (RAVLMImage+Helpers.swift:25-33): maxTokens 256, temperature 0.7,
  // topP 0.9, topK 40. Normalize only when unset/<=0.
  return {
    prompt: options.prompt ?? '',
    maxTokens: options.maxTokens > 0 ? options.maxTokens : 256,
    temperature: options.temperature > 0 ? options.temperature : 0.7,
    topP: options.topP > 0 ? options.topP : 0.9,
    topK: options.topK > 0 ? options.topK : 40,
    stopSequences: options.stopSequences ?? [],
    streamingEnabled,
    systemPrompt: options.systemPrompt,
    maxImageSize: options.maxImageSize ?? 0,
    nThreads: options.nThreads ?? 0,
    useGpu: options.useGpu ?? true,
    modelFamily: options.modelFamily || VLMModelFamily.VLM_MODEL_FAMILY_AUTO,
    customChatTemplate: options.customChatTemplate,
    imageMarkerOverride: options.imageMarkerOverride,
    seed: options.seed ?? 0,
    repetitionPenalty: options.repetitionPenalty > 0 ? options.repetitionPenalty : 1,
    minP: options.minP ?? 0,
    emitImageEmbeddings: options.emitImageEmbeddings ?? false,
  };
}
